src/py/flwr/server/workflow/default_workflows.py

```diff
@@ -0,0 +1,346 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Legacy default workflows."""
+
+
+import timeit
+from logging import DEBUG, INFO
+from typing import Optional, cast
+
+import flwr.common.recordset_compat as compat
+from flwr.common import ConfigsRecord, Context, GetParametersIns, log
+from flwr.common.constant import (
+    MESSAGE_TYPE_EVALUATE,
+    MESSAGE_TYPE_FIT,
+    MESSAGE_TYPE_GET_PARAMETERS,
+)
+
+from ..compat.app_utils import start_update_client_manager_thread
+from ..compat.legacy_context import LegacyContext
+from ..driver import Driver
+from ..typing import Workflow
+
+KEY_CURRENT_ROUND = "current_round"
+KEY_START_TIME = "start_time"
+CONFIGS_RECORD_KEY = "config"
+PARAMS_RECORD_KEY = "parameters"
+
+
+class DefaultWorkflow:
+    """Default workflow in Flower."""
+
+    def __init__(
+        self,
+        fit_workflow: Optional[Workflow] = None,
+        evaluate_workflow: Optional[Workflow] = None,
+    ) -> None:
+        if fit_workflow is None:
+            fit_workflow = default_fit_workflow
+        if evaluate_workflow is None:
+            evaluate_workflow = default_evaluate_workflow
+        self.fit_workflow: Workflow = fit_workflow
+        self.evaluate_workflow: Workflow = evaluate_workflow
+
+    def __call__(self, driver: Driver, context: Context) -> None:
+        """Execute the workflow."""
+        if not isinstance(context, LegacyContext):
+            raise TypeError(
+                f"Expect a LegacyContext, but get {type(context).__name__}."
+            )
+
+        # Start the thread updating nodes
+        thread, f_stop = start_update_client_manager_thread(
+            driver, context.client_manager
+        )
+
+        # Initialize parameters
+        default_init_params_workflow(driver, context)
+
+        # Run federated learning for num_rounds
+        log(INFO, "FL starting")
+        start_time = timeit.default_timer()
+        cfg = ConfigsRecord()
+        cfg[KEY_START_TIME] = start_time
+        context.state.configs_records[CONFIGS_RECORD_KEY] = cfg
+
+        for current_round in range(1, context.config.num_rounds + 1):
+            cfg[KEY_CURRENT_ROUND] = current_round
+
+            # Fit round
+            self.fit_workflow(driver, context)
+
+            # Centralized evaluation
+            default_centralized_evaluation_workflow(driver, context)
+
+            # Evaluate round
+            self.evaluate_workflow(driver, context)
+
+        # Bookkeeping
+        end_time = timeit.default_timer()
+        elapsed = end_time - start_time
+        log(INFO, "FL finished in %s", elapsed)
+
```

Comment:
After this, should we add the logs that are currently in `run_fl`?

Reply:
Yes. I've just added the logs. Btw, may I know what `app_fit` means in the logs?

Reply:
it's just a historic artifact. i think it refers to the fact that this was once placed in the `app` module. we should update it at some point.